core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1514,364 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
+ */
+static void fat_entry_dir_deinit(void)
+{
+	struct rpmb_fat_entry *fe = NULL;
+
+	if (!fat_entry_dir)
+		return;
+
+	if (CFG_RPMB_FS_CACHE_SIZE == 0) {
+		fat_entry_dir_free();
+		return;
+	}
+
+	fe = fat_entry_dir->rpmb_fat_entry_buf;
+	fat_entry_dir->idx_curr = 0;
+	fat_entry_dir->num_total_read = 0;
+	fat_entry_dir->last_reached = false;
+
+	/*
+	 * If the cache is full and we used our temporary buffer, only keep
+	 * cached entries, shrink down to cache size.
+	 */
+	if (fat_entry_dir->num_buffered > CFG_RPMB_FS_CACHE_SIZE) {
+		fat_entry_dir->num_buffered = CFG_RPMB_FS_CACHE_SIZE;
+		/* We reduce the size of the alloc, so this should be good. */
+		fe = realloc(fe, fat_entry_dir->num_buffered * sizeof(*fe));
+
+		/*
+		 * In case realloc fails, we are on the safe side if we destroy
+		 * the whole structure. Upon the next init, the cache has to be
+		 * re-established, but this case should not happen in practice.
+		 */
+		if (!fe)
+			fat_entry_dir_free();
+		else
+			fat_entry_dir->rpmb_fat_entry_buf = fe;
+	}
+}
+
+/**
+ * fat_entry_dir_update: Updates a persisted FAT FS entry in the cache.
+ * This function updates the FAT entry fat_entry that was written to address
+ * fat_address onto RPMB storage in the cache.
+ * Returns TEE_SUCCESS on success.
+ */
+static TEE_Result fat_entry_dir_update(struct rpmb_fat_entry *fat_entry,
+				       uint32_t fat_address)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t fat_entry_buf_idx;
+	/* Use a temp var to avoid compiler warning if caching disabled. */
+	uint32_t max_cache_entries = CFG_RPMB_FS_CACHE_SIZE;
+
+	/* Nothing to update if the cache is not initialized. */
+	if (!fat_entry_dir)
+		goto out;
+
+	/* Sanity check for offset of FAT address. */
+	if ((fat_address - RPMB_FS_FAT_START_ADDRESS) %
+	     sizeof(struct rpmb_fat_entry)) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	/* Determine index in cache. */
+	fat_entry_buf_idx = (fat_address - RPMB_FS_FAT_START_ADDRESS) /
+			     sizeof(struct rpmb_fat_entry);
+
+	/* Only need to write if index points to an entry in cache. */
+	if (fat_entry_buf_idx < max_cache_entries) {
+		memcpy(fat_entry_dir->rpmb_fat_entry_buf + fat_entry_buf_idx,
+		       fat_entry, sizeof(struct rpmb_fat_entry));
+	}
+
+out:
+	return res;
+}
+
+/**
+ * fat_entry_dir_get_next: Get next FAT FS entry.
+ * Read either from cache/buffer, or by reading from RPMB storage if the
+ * elements in the buffer/cache are fully read. When reading in from RPMB
+ * storage, the buffer is overwritten in case caching is disabled.
+ * In case caching is enabled, the cache is either further filled, or a
+ * temporary buffer populated if the cache is already full.
+ * The FAT FS entry is written to fat_entry. The respective address in RPMB
+ * storage is written to fat_address, if not NULL. When the last FAT FS entry
+ * was previously read, the function indicates this case by writing a NULL
+ * pointer to fat_entry.
```

Comment:
Why not return `TEE_ERROR_ITEM_NOT_FOUND` instead? Would simplify the loop below.

Reply:
At first sight I agreed with @jforissier but looking back, it is quite handy that this function does not mix error codes and `TEE_ERROR_ITEM_NOT_FOUND` returned from `tee_rpmb_read()` should maybe be treated as an error case, while 'no next entry' is a valid case.
Therefore I think it is better to return here `TEE_SUCCESS` and a NULL `fat_entry` when not more entries to be read.